File: BOJ/basic-algo-lecture/0x19 트리/4803.py
import sys
input = lambda: sys.stdin.readline().rstrip()
try:
    sys.stdin = open('input.txt', 'r')
except:
    pass

from collections import defaultdict

def dfs(v):
    global cycle
    if vis[v]:
        cycle = True
        return
    vis[v] = 1
    for nv in adj[v]:
        if nv == p[v]:
            continue
        p[nv] = v
        dfs(nv)

def solve():
    global cycle
    ret = 0
    for i in range(1, n + 1):
        if vis[i]:
            continue
        cycle = False
        dfs(i)
        if not cycle:
            ret += 1
    return ret

tc = 1
while True:
    n, m = map(int, input().split())
    if n == m == 0:
        break
    adj = defaultdict(list)
    vis = [0] * (n + 1)
    p = [0] * (n + 1)
    for _ in range(m):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)
    ans = solve()
    print(f"Case {tc}: ", end='')
    if ans == 0:
        print("No trees.")
    elif ans == 1:
        print("There is one tree.")
    else:
        print(f"A forest of {ans} trees.")
    tc += 1

File: BOJ/basic-algo-lecture/0x19 트리/4803.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[505];
bool vis[505];
int p[505];
bool cycle;
int n;

void dfs(int v) {
    if (vis[v]) {
        cycle = true;
        return;
    }
    vis[v] = true;
    for (int nv : adj[v]) {
        if (nv == p[v]) continue;
        p[nv] = v;
        dfs(nv);
    }
}

int solve() {
    int ret = 0;
    for (int i = 1; i <= n; i++) {
        if (vis[i]) continue;
        cycle = false;
        dfs(i);
        if (!cycle) ret++;
    }
    return ret;
}

int main(void) {
    int tc = 1;
    while (true) {
        int m;
        cin >> n >> m;
        if (n == 0 && m == 0) return 0;
        for (int i = 1; i <= n; i++) adj[i].clear();
        fill(vis, vis + n + 1, false);
        fill(p, p + n + 1, 0);
        while (m--) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        int ans = solve();
        cout << "Case " << tc << ": ";
        if (ans == 0) cout << "No trees.\n";
        else if (ans == 1) cout << "There is one tree.\n";
        else cout << "A forest of " << ans << " trees.\n";
        tc++;
    }
}